File: src/main/kotlin/dev/jefy/connectpro/portfolio/applicaion/command/PortfolioCommand.java
package dev.jefy.connectpro.portfolio.applicaion.command;

import org.jspecify.annotations.NullMarked;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

import dev.jefy.connectpro.management.domain.vo.BadgeId;
import dev.jefy.connectpro.portfolio.applicaion.dtos.*;
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId;
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioType;
import dev.jefy.connectpro.portfolio.domain.vo.SocialLinkId;

/**
 * @author Jôph Yamba
 */
@NullMarked
public interface PortfolioCommand {
    PortfolioId create(PortfolioRequest request);
    PortfolioId updateGeneralInfo(PortfolioId portfolioId, GeneralInfoRequest request);
    PortfolioId updateProfessional(PortfolioId portfolioId, ProfessionalInfoRequest request);
    PortfolioId updateContactInfo(PortfolioId portfolioId, ContactInfoData request);
    PortfolioId updateLocationInfo(PortfolioId portfolioId, LocationInfoData request);
    PortfolioId setCoverImage(PortfolioId portfolioId, MultipartFile image) throws IOException;
    PortfolioId deleteCoverImage(PortfolioId portfolioId) throws IOException;
    PortfolioId addSocialLink(PortfolioId portfolioId, SocialLinkData linkData);
    PortfolioId deleteSocialLink(PortfolioId portfolioId, SocialLinkId socialLinkId);
    PortfolioId activate(PortfolioId portfolioId);
    PortfolioId deactivate(PortfolioId portfolioId);
    PortfolioId bloc(PortfolioId portfolioId);
    PortfolioId changeType(PortfolioId portfolioId, PortfolioType type);
    PortfolioId setBadge(PortfolioId portfolioId, BadgeId badgeId);
}

File: src/main/kotlin/dev/jefy/connectpro/portfolio/application/command/PortfolioCommand.kt
package dev.jefy.connectpro.portfolio.application.command

import dev.jefy.connectpro.management.domain.vo.BadgeId
import dev.jefy.connectpro.portfolio.application.dtos.*
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId
import dev.jefy.connectpro.portfolio.domain.vo.SocialPlatform
import org.springframework.web.multipart.MultipartFile
import java.io.IOException

/**
 * @author Jôph Yamba
 */
interface PortfolioCommand {
    fun create(request: CreatePortfolioRequest): PortfolioId
    fun update(portfolioId: PortfolioId, request: UpdatePortfolioRequest,): PortfolioId
    fun updateContact(portfolioId: PortfolioId, request: ContactData): PortfolioId
    fun updateLocation(portfolioId: PortfolioId, request: LocationData): PortfolioId

    @Throws(IOException::class)
    fun setCoverImage(portfolioId: PortfolioId, image: MultipartFile): PortfolioId

    @Throws(IOException::class)
    fun deleteCoverImage(portfolioId: PortfolioId): PortfolioId
    fun addSocialLink(portfolioId: PortfolioId, social: SocialData): PortfolioId
    fun deleteSocialLink(portfolioId: PortfolioId, platform: SocialPlatform): PortfolioId
    fun activate(portfolioId: PortfolioId): PortfolioId
    fun deactivate(portfolioId: PortfolioId): PortfolioId
    fun bloc(portfolioId: PortfolioId): PortfolioId
    fun setBadge(portfolioId: PortfolioId, badgeId: BadgeId): PortfolioId
}
